File: TP01/TP01Q09/Arquivo.c
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q09 - 13 / 03 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : gcc -o Arquivo Arquivo.c
 *       Windows: gcc -o Arquivo Arquivo.c
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : ./Arquivo
 *       Windows: Arquivo
 *   
*/

// ---------------------------------------- Dependencias

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// ---------------------------------------- Metodos

/**
 *  Funcao para escrever numero real no arquivo.
 *  @param input - double: Numero a ser escrito.
 *  @param filename - char*: Nome do arquivo.
*/
void writeDoubleToFile( double input, char* filename )
{
    FILE *file = fopen( filename, "ab" );
    fwrite( &input, sizeof(double), 1, file );
    fclose( file );
} // end writeDoubleToFile ( )

/**
 *  Funcao para formatar um numero real e printar para o Verde.
 *  Logica: Converter de double para char e manipular os caracteres. Se for '0' substituir por '\0' e se for '.' o
 *  numero e' inteiro, entao tirar o '.' tambem.
 *  @param input - double: Numero real a ser formatado.
*/
void formattedDouble( double input )
{
    char *buffer = (char*) malloc( (80+1) * sizeof( char ) ); 
    sprintf( buffer, "%.10lf", input );
    int length = strlen( buffer );
    while( length > 0 && ( buffer[length-1] == '0' || buffer[length-1] == '.' ) )
    {
        if( buffer[length-1] == '.' )
        {
            buffer[length-1] = '\0';
            length = 0;
        } // end if
        buffer[length-1] = '\0'; 
        length = length - 1;
    } // end while
    printf( "%s\n", buffer );
} // end formatDouble ( )

/**
 *  Funcao para ler numero real do arquivo.
 *  @param n - int: Quantidade de numeros no arquivo.
 *  @param filename - char*: Nome do arquivo.
*/
void readDoubleFromFile( int n, char* filename )
{
    FILE *file = fopen( filename, "rb" );
    double value = 0.0;
    if( file == NULL )
    {
        printf( "\n%s\n", "ERRO: Nao foi possivel abrir o arquivo" );
    }
    else
    {
        for( int pointer = 0; pointer < n; pointer = pointer + 1 )
        {
            fseek( file, (n - pointer - 1) * sizeof(double) , SEEK_SET);
            fread( &value, sizeof(double), 1 ,file );
            formattedDouble( value );
        } // end for
    } // end if
    fclose( file );
} // end readDoubleFromFile ( )

int main ( int argc, char* argv[] )
{
    int n = 0;
    double input = 0.0;
    char *filename = "ARQUIVO.TXT";

    scanf( "%d", &n );
    
    for( int x = 0; x < n; x = x + 1 )
    {
        scanf( "%lf", &input );
        writeDoubleToFile( input, filename );
    } // end for

    readDoubleFromFile( n, filename );
} // end main ( )

File: TP01/TP01Q08/Arquivo.java
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q08 - 13 / 03 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : javac Arquivo.java
 *       Windows: javac Arquivo.java
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : java Arquivo
 *       Windows: java Arquivo
 *   
*/

import java.io.RandomAccessFile;

public class Arquivo 
{
    /**
     * Quantidade de numeros reais.
     */
    private static int n = 0;

    /**
     *  Funcao principal
     *  @param args
     */
    public static void main ( String[] args )
    {
        MyIO.setCharset("UTF-8");
        double input = 0.0;
        String filename = "ARQUIVO.TXT";

        n = MyIO.readInt();

        for( int x = 0; x < n; x = x + 1 )
        {
            input = MyIO.readDouble();
            writeDoubleToFile(input, filename);
        } // end for

        readDoubleFromFile( filename );
    } // end main ( )

    /**
     *  Funcao para escrever numero real no arquivo.
     *  @param input - Double: Numero real a ser escrito no arquivo.
     *  @param filename - String: Nome do arquivo.
     */
    public static void writeDoubleToFile( double input, String filename )
    {
        try 
        {
            RandomAccessFile file = new RandomAccessFile( filename, "rw" );
            file.seek( file.length() );
            file.writeDouble( input );
            file.close();
        } 
        catch( Exception e )
        {
            MyIO.println( "ERRO: " + e.getMessage() );
        } // end try catch
    } // end writeDoubleToFile ( )

    /**
     *  Funcao para ler numero real do arquivo e mostrar na tela.
     *  @param filename - String: Nome do Arquivo.
     */
    public static void readDoubleFromFile( String filename )
    {
        double value = 0.0;
        try 
        {
            RandomAccessFile file = new RandomAccessFile( filename, "r" );
            for( int pointer = (int)file.length()-8; pointer >= 0; pointer = pointer - 8 )
            {
                file.seek(pointer);
                value = file.readDouble();
                if( value - (int)value == 0 )
                {  
                    MyIO.println( (int)value );
                }
                else{
                    MyIO.println( value );
                } // end if
            } // end for
            file.close();
        } 
        catch( Exception e )
        {
            MyIO.println( "ERRO: " + e.getMessage() );
        } // end try catch
    } // end readDoubleToFile ( )

} // end class
